dowhy/causal_refuters/dummy_outcome_refuter.py

```diff
@@ -160,26 +167,86 @@ def refute_estimate(self):
+            elif action == 'noise':
+                function_pipeline.append( ( lambda new_data: self._noise(new_data, func_args), {'input':['y']} ) )
+            elif action == 'permute':
+                function_pipeline.append( ( lambda new_data: self._permute(new_data, func_args), {'input':['y']} ) )
+            elif action == 'zero':
+                function_pipeline.append( (lambda new_data: np.zeros(new_data.shape), {'input':['y']} ) )
+            else:
+                raise ValueError("Invalid action:{} passed to pipeline".format(action))
 
-    def _run_estimation_function(self):
-        estimator = self._get_regressor_object()
+        return function_pipeline
+            
+    def _estimate_dummy_outcome(self, func_args, action, new_data):
+        pdb.set_trace()
+        estimator = self._get_regressor_object(action, func_args)
         X = self._data[self._chosen_variables]
-        y = self._data['y']
+        y = new_data
         estimator = estimator.fit(X, y)
         
-        return estimator.predict(X) 
+        return estimator.predict
     
-    def _get_regressor_object(self):
+    def _get_regressor_object(self, action, func_args):
+        if  action == "linear_regression":
+            return LinearRegression(**func_args)
+        elif action == "knn":
+            return KNeighborsRegressor(**func_args)
+        elif action == "svm":
+            return SVR(**func_args)
+        elif action == "random_forest":
+            return RandomForestRegressor(**func_args)
+        elif action == "neural_network":
+            return MLPRegressor(**func_args)
+        else:
+            raise ValueError("The function: {} is not supported by dowhy at the moment".format(action))
+
+    def _permute(self, new_data, permute_fraction):
+        new_data = pd.DataFrame(new_data)
+        new_data.columns = ['y']
+
+        if permute_fraction == 1:
+            return new_data['y'].sample(frac=1).values
+        else: 
+            return new_data.apply( lambda row: self.replace(row, new_data, permute_fraction), axis=1).values
         
-        if  self._outcome_function == "linear_regression":
-            return LinearRegression(**self._params)
-        elif self._outcome_function == "knn":
-            return KNeighborsRegressor(**self._params)
-        elif self._outcome_function == "svm":
-            return SVR(**self._params)
-        elif self._outcome_function == "random_forest":
-            return RandomForestRegressor(**self._params)
-        elif self._outcome_function == "neural_network":
-            return MLPRegressor(**self._params)
+
+    def replace(self, row, new_data, fraction):
+        '''
+        In this function, we make use of the Fisher Yates shuffle:
+        https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle
+        '''  
+        if np.random.uniform(0,1) <= fraction:

```

Comment:
I don't understand this method. Can't this lead to duplicates---one row being twice in the final data?
A simpler way is to first sample f rows randomly. where f is the fraction using code on line 244. Let the indices be ind_arr
then, new_data[ind_arr] = new_data[ind_arr.sample(frac=1)] should work.